src/renderer/atlas/BackendD3D.cpp

```diff
@@ -1025,58 +995,53 @@ void BackendD3D::_drawText(RenderingPayload& p)
         for (const auto& m : row->mappings)
         {
             auto x = m.glyphsFrom;
-            const AtlasFontFaceKey fontFaceKey{
-                .fontFace = m.fontFace.get(),
-                .lineRendition = row->lineRendition,
-            };
+            const auto glyphsTo = m.glyphsTo;
+            const auto fontFace = m.fontFace.get();
 
-        // This goto label exists to allow us to retry rendering a glyph if the glyph atlas was full.
-        // We need to goto here, because a retry will cause the atlas texture as well as the
-        // _glyphCache hashmap to be cleared, and so we'll have to call insert() again.
-        drawGlyphRetry:
-            const auto [fontFaceEntryOuter, fontFaceInserted] = _glyphAtlasMap.insert(fontFaceKey);
-            auto& fontFaceEntry = *fontFaceEntryOuter.inner;
-
-            if (fontFaceInserted)
+            // The lack of a fontFace indicates a soft font.
+            AtlasFontFaceEntry* fontFaceEntry = &_builtinGlyphs;
+            if (fontFace) [[likely]]
             {
-                _initializeFontFaceEntry(fontFaceEntry);
+                fontFaceEntry = _glyphAtlasMap.insert(fontFace).first;
             }
 
-            while (x < m.glyphsTo)
+            const auto& glyphs = fontFaceEntry->glyphs[WI_EnumValue(row->lineRendition)];
+
+            while (x < glyphsTo)
             {
-                const auto [glyphEntry, inserted] = fontFaceEntry.glyphs.insert(row->glyphIndices[x]);
+                u32 glyphIndex = row->glyphIndices[x];
+                if (!fontFace && til::is_leading_surrogate(glyphIndex))
+                {
+                    glyphIndex = til::combine_surrogates(glyphIndex, row->glyphIndices[x + 1]);

```

Comment:
Just confirming - It is not possible for us to confuse these codepoints with actual glyph indices mapped from any font, right? Because glyph indices are tracked per font face entry?

Reply:
Yup! Due to the `if (!fontFace && ...` prefix this entire condition is skipped for all regular font mappings.